benches/aead.rs

```diff
@@ -0,0 +1,186 @@
+    fn advance(&mut self) -> Result<aead::Nonce, error::Unspecified> {
+        let mut result = [0u8; aead::NONCE_LEN];
+        result[4..].copy_from_slice(&u64::to_be_bytes(self.0));
+        self.0 = self.0.checked_add(1).ok_or(error::Unspecified)?;
+        Ok(aead::Nonce::assume_unique_for_key(result))
+    }
+}
+
+macro_rules! function_bench_name {
+    ( $benchmark_name:ident, $algorithm:expr, $operation:ident) => {{
+        const FUNC_NAME_BASE: &str = concat!(
+            stringify!($operation),
+            "_",
+            stringify!($benchmark_name),
+            "_",
+            stringify!($algorithm)
+        );
+
+        FUNC_NAME_BASE.replace("&aead::", "")
+    }};
+}
+
+macro_rules! bench {
+    ( $benchmark_name:ident, $algorithm:expr, $chunk_len:expr, $ad:expr ) => {
+        pub(super) fn $benchmark_name(c: &mut criterion::Criterion) {
+            use super::{NonceSequence, NONCE};
+            use criterion::{black_box, BatchSize};
+            use ring::{
+                aead::{self, BoundKey},
+                rand::{SecureRandom, SystemRandom},
+            };
+
+            let rng = SystemRandom::new();
+
+            let mut key_bytes = vec![0u8; $algorithm.key_len()];
+            rng.fill(&mut key_bytes).unwrap();
+
+            {
+                let key = aead::UnboundKey::new($algorithm, &key_bytes).unwrap();
+                let mut key = aead::SealingKey::new(key, NonceSequence::new());
+
+                c.bench_function(
+                    &function_bench_name!($benchmark_name, $algorithm, encrypt),
+                    |b| {
+                        b.iter_batched(
+                            || black_box(vec![0u8; $chunk_len + $algorithm.tag_len()]),
+                            |mut in_out| -> Result<(), ring::error::Unspecified> {
+                                let aad = aead::Aad::from(black_box($ad));
+                                let _result = key.seal_in_place_append_tag(aad, &mut in_out)?;
+                                Ok(())
+                            },
+                            BatchSize::SmallInput,
+                        )
+                    },
+                );
+            }
+
+            {
+                let key =
+                    aead::LessSafeKey::new(aead::UnboundKey::new($algorithm, &key_bytes).unwrap());
+
+                let ciphertext = {
+                    let nonce = aead::Nonce::assume_unique_for_key(NONCE);
+                    let mut in_out = vec![0u8; $chunk_len + $algorithm.tag_len()];
+                    let aad = aead::Aad::from($ad);
+                    key.seal_in_place_append_tag(nonce, aad, &mut in_out)
+                        .unwrap();
+                    in_out
+                };
+
+                c.bench_function(
+                    &function_bench_name!($benchmark_name, $algorithm, decrypt),
+                    |b| {
+                        b.iter_batched(
+                            || black_box(ciphertext.clone()),
```

Comment:
I've moved this to use `iter` and a ciphertext reset at the end instead of making a new allocation/deallocation every time.

Note that I've extracted the reset into its own function in order for Rust to see its just a `memcpy` and save a lot of instructions. For some reason, it didn't see the optimization on the `Vec`s themselves, or after coercing them to slices in the same function. You can try this out [here](https://rust.godbolt.org/z/Ke41hhrTo) if you'd like.